File: 142.Linked List Cycle II/142.Linked List Cycle  II.java
题目要求是判断链表是否有环，leetcode上链表的题都是没有头结点的，这点大家要记住。而且若链表有环，也是最后一个节点形成的环。

大家考虑这样一个问题，链表的环相当于一个圆形操场。假设有两个人在圆形操场上无限循环的跑，那么速度快的一定能追得上速度慢的。
同理，若要判断一个链表是否有环，可设计快慢指针，当快慢指针都进入环的时候，若最终两个指针相遇，必可说明链表存在环。
下面就要考虑快慢指针的步长，从跑操场的情况来看，不管慢的有多慢，快得有多快，最终肯定能相遇。同理，链表中，也可随意指定快慢指针的步长，无非就是跑的圈数多少的问题

 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode p=head;//快指针
        ListNode q=head;//慢指针
        while(p!=null&&q!=null&&p.next!=null){//边界条件是出现空指针，就返回false；
            q=q.next;
            p=p.next.next;//空指针没有next，否则会出现NullPointerException问题
            if(p==q)return true;
        }
        return false;
    }
}

reference：
        https://blog.csdn.net/shuaishuai3409/article/details/51434182

File: 142.Linked List Cycle II/142.Linked List Cycle  II.cpp
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_map<ListNode*, bool> visited;
        while(head != NULL)
        {
            if(visited[head] == true)
                return head;
            visited[head] = true;
            head = head->next;
        }
        return NULL;
    }
};
解法二：
使用快慢指针，
fast每次前进两步(若两步中出现NULL，则返回NULL).
slow每次前进一步(若出现NULL，则返回NULL).
当出现环路，则总有某时刻，fast会赶上slow（相遇），证明如下：
1、若fast套slow圈之前在slow后方两步，则fast2slow1之后，fast在slow后方一步，进入(2)
2、若fast套slow圈之前在slow后方一步，则fast2slow1之后，fast追上slow（相遇）
其他情况均可化归到以上两步。
假设从head到环入口entry步数为x，环路长度为y，相遇时离entry的距离为m
则fast：x+ay+m，slow：x+by+m  (a > b)
x+ay+m = 2(x+by+m)
整理得x+m = (a-2b)y
以上式子的含义是，相遇时的位置(m)再前进x步，正好是y的整倍数即整圈。
现在的问题是怎样计数x。
利用head到entry的长度为x，只要fast从head每次前进一步，slow从相遇位置每次前进一步，
再次相遇的时候即环的入口。

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;
        do
        {
            if(fast != NULL)
                fast = fast->next;
            else
                return NULL;
            if(fast != NULL)
                fast = fast->next;
            else
                return NULL;
            slow = slow->next;
        }while(fast != slow);
        fast = head;
        while(fast != slow)
        {
            fast = fast->next;
            slow = slow->next;
        }
        return slow;
    }
};